Boost: Support background-images for LCP Optimization

LiamSarsfield opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes [HOG-69: Boost Plugin: Support background-image](https://linear.app/a8c/issue/HOG-69/boost-plugin-support-background-image)

## Proposed changes:

*   Adds preload links (`<link rel="preload" as="image" fetchpriority="high">`) to the `<head>` for background images identified as the Largest Contentful Paint (LCP) element. This helps browsers discover and load these critical images earlier.
*   Retrieves LCP background image data stored by `LCP_Storage` for the current request.
*   Adds a new action hook `add_preload_links_to_head` on `wp_head` to inject the preload links.
*   Refactors the conditions for skipping LCP optimization into a new private method `should_skip_optimization()` for clarity and reuse.
*   Updates the `optimize()` method to handle potential 'unknown' LCP types gracefully and ensures it only attempts image optimization logic for elements explicitly identified as `img`.


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
pc9hqz-3ul-p2

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
N/A

## Testing instructions:

1.  **Setup:**
    *   Ensure Jetpack Boost plugin is installed and activated.
    *   **Tip:** To make testing easier, you can temporarily force a specific element to likely be the LCP. Add one of the following PHP snippets to your theme's `functions.php` file or a custom plugin. Remember to remove it after testing. *Note: These snippets increase the likelihood but don't guarantee LCP status; always verify with browser dev tools.*

        <details>
        <summary>Click here for snippet to force a background image LCP</summary>

        ```php
        // Add CSS to wp_head to style the element
        add_action( 'wp_head', function() {
            // Replace with a URL to a large image on your site
            $image_url = 'URL_TO_YOUR_LARGE_BACKGROUND_IMAGE.jpg';
            ?>
            <style>
              #force-lcp-background {
                width: 100%;
                height: 80vh; /* Adjust height as needed */
                background-image: url('<?php echo esc_url( $image_url ); ?>');
                background-size: cover;
                background-position: center;
              }
            </style>
            <?php
        } );

        // Add the element early in the body
        add_action( 'wp_body_open', function() {
            echo '<div id="force-lcp-background"></div>';
        } );
        ```
        *(Remember to replace `URL_TO_YOUR_LARGE_BACKGROUND_IMAGE.jpg`)*
        </details>

        <details>
        <summary>Click here for snippet to force a standard <img> LCP</summary>

        ```php
        // Add the image early in the body
        add_action( 'wp_body_open', function() {
             // Replace with a URL to a large image on your site
            $image_url = 'URL_TO_YOUR_LARGE_IMAGE.jpg';
             // Adjust width/height as needed to make it the largest element
            echo '<img src="' . esc_url( $image_url ) . '" width="800" height="600" alt="Forced LCP Image" style="display:block; margin: 20px auto;">';
        } );
        ```
        *(Remember to replace `URL_TO_YOUR_LARGE_IMAGE.jpg` and adjust dimensions if needed)*
        </details>

2.  **Test Case 1: LCP Background Image**
    *   Use the PHP snippet above (or find a page naturally) where the LCP element is a CSS background image.
    *   Enable the LCP optimization feature within Boost settings, and wait for optimization to complete
    *   Visit this page on the frontend.
    *   View the page source (HTML).
    *   **Verify:** In the `<head>` section, check for a `<link rel="preload" href="[URL_of_LCP_background_image]" as="image" fetchpriority="high" />` tag. The URL should match the background image identified as the LCP element.

3.  **Test Case 2: LCP Standard Image**
    *   Use the PHP snippet above (or find a page naturally) where the LCP element is a standard `<img>` tag.
    *   Enable the LCP optimization feature within Boost settings, and wait for optimization to complete
    *   Visit this page on the frontend.
    *   View the page source.
    *   **Verify:** Check that the standard LCP `<img>` tag still has the `fetchpriority="high"` attribute added correctly by the existing optimization logic.
    *   **Verify:** Ensure no *additional* preload link for this `<img>` tag is added to the `<head>` by the *new* background image logic.

LiamSarsfield committed:
Enhance LCP Optimization: Add Preload Links for Background Images

- Introduced a new method to add preload links for LCP background images in the <head> section.
- Implemented checks to skip optimization based on various conditions.
- Updated the optimization logic to handle image tags more effectively.

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Merge branch 'trunk' into add/lcp/bg-img-support-plugin

haqadn commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -257,6 +318,10 @@ private function optimize_viewport( $buffer, $lcp_data ) {
+		if ( $lcp_data['type'] === 'unknown' ) {
+			return $buffer;
+		}
+
```
Instead of this, it might be beneficial to check the type against a known list of types that work. This would maintain compatibility in case a new type is added on the cloud but the plugin version is lagging behind. For example, we add support for a new type in the cloud, let's call it, "youtube-embed". Then we release a plugin update to support that type. However, sites with older version of the plugin will continue to work with the result if we take the approach below.

```suggestion
		if ( in_array( $lcp_data['type'], array( 'background-image', 'img' ), 1 ) {
			return $buffer;
		}

```

LiamSarsfield replied:
Good point, I've updated it now to return early if it's not a type the plugin knows how to handle.

haqadn commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -265,15 +330,14 @@ private function optimize_viewport( $buffer, $lcp_data ) {
 
-		// Create the optimized tag with required attributes.
-		$optimized_tag = $this->optimize_image_tag( $lcp_html );
+		if ( $lcp_data['type'] === 'img' ) {
```
The types should be moved to constants to avoid mistakes while using.

haqadn commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -130,71 +191,71 @@ public function start_output_filtering() {
 		if ( ! ( new LCP_State() )->is_analyzed() ) {
-			return;
+			return true;
 		}
```
It's best if we do not involve a state check while optimizing the images. Instead, it should check if an optimization for the current page is available or not.

Why? Because, I am seeing the state more of a helper for tracking how the analysis is going rather than being part of the optimization itself. We may or may not want to stop optimizing while an analysis is ongoing. If we do not want to show the old optimizations, we can simply clear the storage of any prior optimizations.

LiamSarsfield replied:
That's a fair explanation, I've updated it to instead return early if there is no LCP storage for the current request.

haqadn commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -130,71 +191,71 @@ public function start_output_filtering() {
 		if ( ! Cornerstone_Utils::is_current_page_cornerstone() ) {
-			return;
+			return true;
 		}
```
It's a nitpick. We should not worry about a page being a cornerstone while applying the optimization. If there is somehow data in the storage for the current page, let's display it.

My reasoning is to remove any unnecessary roadblocks when we do decide to expand the scope of this feature or make it easy if we are running a unit test.

LiamSarsfield replied:
> If there is somehow data in the storage for the current page, let's display it.

Yeah this makes sense, it won't do any harm if this condition is somehow not met currently.

haqadn commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -130,71 +191,71 @@ public function start_output_filtering() {
 		if ( false === apply_filters( 'jetpack_boost_should_optimize_lcp', true ) ) {
-			return;
+			return true;
 		}
```
In conjunction with my suggestion to remove the cornerstone check and `is_analyzed` check, I would suggest you move this filter outside of this function and instead run the output of this function through this filter. That would make it more powerful and let you not only force disable optimization, but also to force enable it if needed.

For example:
```php
$skipping_lcp = apply_filters( 'jetpack_boost_should_optimize_lcp', $this->should_skip_optimization() );
```

LiamSarsfield replied:
I like this pattern! I've applied this to both `start_output_filtering`, and the `add_preload_links_to_head` functions.

dilirity replied:
Since @haqadn made the suggestion initially, I'll leave a note here as well for my suggestion in https://github.com/Automattic/jetpack/pull/43150#discussion_r2073167186

haqadn reviewed (dismissed):
I have suggested a few changes. Feel free to reject as you see fit, I understand many of them might be out of scope for this PR.

One thing that is most likely OOS but would be really nice would be to extract the optimization of each type to their own class. This would give us a nice way to put relevent code untangled and keep the main class lite. I was thinking something like:

```php
// Create an optimizer from a factory function that returns `new Background_Image_Optimizer( $lcp_data['element'] )` or `new Img_Tag_Optimizer( $lcp_data['element'] )`
$optimizer = $this->create_optimizer_from_data( $lcp_data );
$optimizer->optimize( $buffer );
```

dilirity commented:
@haqadn 

> One thing that is most likely OOS but would be really nice would be to extract the optimization of each type to their own class. This would give us a nice way to put relevent code untangled and keep the main class lite. I was thinking something like:

Just a note, I see @donnchawp has already started work on something in this direction with https://github.com/Automattic/jetpack/pull/43190/

LiamSarsfield committed:
Refactor LCP Optimization Logic and Introduce Constants for Image Types

- Added constants for LCP types: background images and standard images.
- Updated optimization methods to utilize the new constants for better readability and maintainability.
- Enhanced the logic to apply filters for skipping optimization checks.
- Improved handling of LCP data retrieval and processing.

LiamSarsfield commented:
@haqadn Thanks for the review. I've applied all your feedback besides the one that @dilirity [mentioned](https://github.com/Automattic/jetpack/pull/43150#issuecomment-2821417028).
To reduce conflicts, I think this can be done within either #43190, or as a separate effort/PR.

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into add/lcp/bg-img-support-plugin

dilirity commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -122,79 +128,126 @@ public function register_data_sync( $instance ) {
-		 * @since   3.13.1
-		 */
-		if ( false === apply_filters( 'jetpack_boost_should_optimize_lcp', true ) ) {
+		if ( apply_filters( 'jetpack_boost_should_optimize_lcp', $this->should_skip_optimization() ) ) {
```
Filter comment disappeared here.

LiamSarsfield replied:
I'm not sure how that disappeared, but I've added it back now and put a new `$$next-version$$` as the behaviour as slightly changed.

dilirity commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -122,79 +128,126 @@ public function register_data_sync( $instance ) {
+	 * @since $$next-version$$
+	 */
+	public function add_preload_links_to_head() {
+		if ( apply_filters( 'jetpack_boost_should_optimize_lcp', $this->should_skip_optimization() ) ) {
```
Can this go in a separate method? It's already present in `start_output_filtering`.

LiamSarsfield replied:
I've created a new function for this specifically `should_skip_optimization_with_filter`, let me know your thoughts on this.

dilirity replied:
Thanks! I took another look at the whole thing and I got another idea (sorry 😅). Tell me what you think about it.

Currently, the filter is applied after the whole logic in `should_skip_optimization` is executed. What if we add a filter to short circuit that, right at the top of `should_skip_optimization`? For example, take a look at how `pre_get_shortlink` is handled [here](https://developer.wordpress.org/reference/hooks/pre_get_shortlink/).

[Here's the full comment](https://github.com/WordPress/wordpress-develop/blob/6.8.1/src/wp-includes/link-template.php#L4145-L4158).

To further the point, having `should_skip_optimization` and `should_skip_optimization_with_filter` means that you'd be able to use a method that bypasses the filter. Consolidating it in one `should_skip_optimization` with a short circuit filter makes more sense to me.

dilirity replied:
I went ahead and implemented the solution [here](https://github.com/Automattic/jetpack/pull/43150/commits/7a71a9a6aa68f9072f2d3e0a894b7302cd3506f5).

You can try it out with this snippet:

```php
add_filter( 'jetpack_boost_pre_should_skip_lcp_optimization', '__return_true' );
```

LiamSarsfield replied:
The intention of the `should_skip_optimization_with_filter` was so folks could use the returned value of `should_skip_optimization` if they needed it. But with `jetpack_boost_pre_should_skip_lcp_optimization` folks leveraging this filter may not have this data is my understanding. I think it's more beneficial for folks to see the returned value first, then decide on if it should still be. 
But since this is already applied, I'm happy with the current approach.

dilirity commented on projects/plugins/boost/app/modules/optimizations/lcp/class-lcp.php:
```diff
@@ -122,79 +128,126 @@ public function register_data_sync( $instance ) {
 		}
 
-		if ( ! ( new LCP_State() )->is_analyzed() ) {
+		$storage     = new LCP_Storage();
```
If we instantiate the storage in the constructor (like cloud css), we would only need to do it once.

LiamSarsfield replied:
Is this only for consistency or is there performance overheard for instantiating multiple times? Just curious, I agree it's better that its consistent regardless.

dilirity replied:
It's more about the instantiating. But at the same time, we do create instances only once (or at least we should), so sort of for consistency.

Ideally, this shouldn't matter if creating a new instance doesn't create any side-effects (I don't think we have much of those in the code-base).

However, if we were playing around with state and changing it, instantiating it multiple times might lead to inconsistencies across the places it's used.

LiamSarsfield replied:
Sounds good, yes constructors shouldn't create side-effects.

dilirity reviewed (changes requested):
Nice work!

A few small notes and ~we can merge this 🥳~ I'll test this so we can hopefully merge it!

LiamSarsfield committed:
Add a new method to apply filters for skipping optimization

dilirity committed:
Merge branch 'trunk' into add/lcp/bg-img-support-plugin

dilirity committed:
Remove storage check from skip method

dilirity committed:
Add short circuit filter to allow skipping LCP optimization earlier

dilirity reviewed (approved):
Thanks for taking care of this @LiamSarsfield <3

I've made some changes to clean things up.

I think we can merge this.

:shipit:
